### PR TITLE
Autogenerate MITER ATT&CK tags - Issue #3

### PR DESCRIPTION
**Pull request by azureknight63**
This completes issue #3
- Stable mechanism for pulling tags: generate_tags.py parses a JSON file produced by MITRE
- generate_tags.py reorganizes the data into the requested format
- generate_tags.py then outputs a new attack.py and (optionally) a new `__init__.py` file

Located a JSON file of Mitre ATT&CK techniques for the latest version (14).

Composed generate_tags.py to parse the JSON file, extracting useful technique data, and exporting that formatted data into an auto-generated attack.py file.

Replaced the manually-generated attack.py file in the repository with the auto-generated file.

Added generate_init.py to pull the variable names from attack.py and generate a new `__init__.py` file.

You must obtain a copy of enterprise-attack.json from Mitre in order to run generate_tags.py. See comments in generate_tags.py for more information.

Direct any Python interpreter to execute generate_tags.py to produce another attack.py file if any changes become necessary to enterprise-attack.json or to the configurable parameters in generate_tags.py.

### CODE REVIEW

File: monkeyevents/tags/generate_init.py
Comment (mssalvatore) [inline monkeyevents/tags/generate_init.py:1]: This is cool! Very clever.

With such a large number of tags, it's probably better to just modify `__init__.py` to have `from .attack import *`. That would shorten up `__init__.py` and remove the need for this code.

File: monkeyevents/tags/generate_tags.py
Comment (mssalvatore) [inline monkeyevents/tags/generate_tags.py:1]: This file is a utility and should be at the top level of the repository, not inside the monkeyevents package. We don't want to distribute this to users. It's used by developers during development to refresh the `attack.py` file.

File: monkeyevents/tags/generate_tags.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import json
+
+from generate_init import generate_init
+
+# This script loads the enterprise-attack.json mitre dataset into a dict (mitre_dataset)
+# according to the parameters set below. The dict is then iterated to produce an attack.py
+# file, containing variables for the multiple Mitre ATT&CK tags relevant to the project.
+#
+# The enterprise-attack.json file is sourced from here:
+# https://raw.githubusercontent.com/mitre-attack/attack-stix-data/master/enterprise-attack/enterprise-attack.json
```
Comment (mssalvatore) [inline monkeyevents/tags/generate_tags.py:10]: Is there any way to pin this to a specific version?

**Review by mssalvatore — CHANGES_REQUESTED** [review]
This is off to a great start! Thanks for the contribution. See comments below.

File: monkeyevents/tags/generate_init.py
Comment (azureknight63) [reply monkeyevents/tags/generate_init.py:1]: The only issue with this is that it was flagging Flake8 https://www.flake8rules.com/rules/F403.html

Comment (mssalvatore) [reply monkeyevents/tags/generate_init.py:1]: Everything has tradeoffs. Their rational is "`from module import *` is discouraged because the programmer often won’t know where an imported object is defined." I'm going to wager developers can figure out that the attack tags are defined in `tags/attack.py` :p

The drawbacks of not importing `*` IMHO outweigh the benefits of making the imports explicit. The solution is `from .attack import *  # noqa: F403`. That will tell flake8 to ignore F403 warnings on that line.

File: generate_tags.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        response = requests.get(url)
+        if response.status_code == 200:
+            print("\nDownload finished!")
+            return json.loads(response.content)
+        else:
+            print("Error: ", response.status_code)
+            return None
+    except requests.RequestException as e:
+        print("Error: ", e)
+        return None
```
Comment (mssalvatore) [inline generate_tags.py:41]: This function mixes how errors are returned. On the one hand, come errors are caught, such as `RequestException` and `None` is returned as a sentinel value to indicate an error has occurred. On the other, exceptions, such as `JSONDecodeError` are not handled here and are therefore raised to the caller.

1. Pick a single way to communicate errors to the caller. Either return `None` meaning, "I couldn't retrieve the data" or allow errors to be raised back to the caller.
2. Unless the function can take some action based on the error, it's best to raise it back to the caller. It's fine if you want to catch it and add some meaningful context in a print/log statement, but if the error won't cause the function to take some different action, there's not point in catching it here. Catch it somewhere where it's appropriate to make a decision about how to proceed. Or don't and let the program crash.

File: generate_tags.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            ]
+        )
+        .replace("-", "_")
+        .replace(".", "_")
+        .replace("/", "_")
+        .replace("(", "")
+        .replace(")", "")
+        .replace("Ä", "A")
+    )
+    return output_reformatted_technique
```
Comment (mssalvatore) [inline generate_tags.py:66]: Just a nit to pick:
```suggestion
    return (
        "_".join(
            [
                target_technique["name"].upper().replace(" ", "_"),
                target_technique["external_references"][0]["external_id"].upper(),
            ]
        )
        .replace("-", "_")
        .replace(".", "_")
        .replace("/", "_")
        .replace("(", "")
        .replace(")", "")
        .replace("Ä", "A")
    )
```

File: generate_tags.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    "https://raw.githubusercontent.com/mitre-attack/attack-stix-data/"
+    "master/enterprise-attack/enterprise-attack-14.1.json"
+)
+
+# Begin parameters; modify these to customize the tags produced by the script
+
+included_platforms = ["Windows", "Linux"]
+include_subtechniques = True
+include_deprecated = False  # include techniques marked as "deprecated"
+include_revoked = False  # include techniques marked as "revoked"
```
Comment (mssalvatore) [inline generate_tags.py:24]: For the foreseeable future, we'll always want these options set how they are. Let's remove these options and simplify the code.

File: monkeyevents/tags/attack.py
Comment (mssalvatore) [inline monkeyevents/tags/attack.py:1]: The changes made to this file need to be in their own commit, not mixed in with the commit that adds the generation script.

File: generate_tags.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        if technique.get("type") == "attack-pattern":
+            x_mitre_platforms = technique.get("x_mitre_platforms", [])
+            x_mitre_is_subtechnique = technique.get("x_mitre_is_subtechnique", False)
+            x_mitre_deprecated = technique.get("x_mitre_deprecated", False)
+            revoked = technique.get("revoked", False)
+
+            checks = {
+                "platforms": any(platform in x_mitre_platforms for platform in included_platforms),
+                "subtechnique": x_mitre_is_subtechnique == include_subtechniques
+                or not include_subtechniques,
```
Comment (mssalvatore) [inline generate_tags.py:82]: As written, I think this logic filters out techniques. In other words, only subtechniques are included. For example, T1548.001-T1548.005 would be included, but T1548 itself would be missing from the generated `attack.py` file.

**Review by mssalvatore — CHANGES_REQUESTED** [review]
image
The first commit summary here is too long. Commit summaries should be no longer than 72 characters, and preferably under 50.

The second one should be in the imperative mood, e.g. "Add Mitre ATT&CK tag auto-generator". Please read https://cbea.ms/git-commit/ to learn about writing good commit messages.

File: generate_tags.py (hunk 5 above)
Comment (azureknight63) [reply generate_tags.py:82]: Simplifying the options and removing filters for subtechniques, deprecated techniques, and revoked techniques makes this change request inapplicable.

File: generate_tags.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def retrieve_mitre_data(url):
+    print("Attempting download of Mitre data, please wait...")
+    response = requests.get(url)
+    if response.status_code == 200:
+        print("\nDownload finished!")
+    else:
+        print("Error: ", response.status_code)
+    return json.loads(response.content)
```
Comment (mssalvatore) [inline generate_tags.py:33]: See https://3.python-requests.org/api/?highlight=raise_for_status#requests.Response.raise_for_status
```suggestion
    response.raise_for_status()
    
    print("\nDownload finished!")
    return json.loads(response.content)
```

If you do this, you don't need `if mitre_dataset` on line 61.

File: generate_tags.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+if mitre_dataset:
+    valid_techniques = []
+    for technique in mitre_dataset.get("objects", []):
+        if (
+                technique.get("type") == "attack-pattern" and
+                not technique.get("x_mitre_deprecated") and
+                not technique.get("revoked")
+        ):
+            x_mitre_platforms = technique.get("x_mitre_platforms", [])
+            if any(platform in x_mitre_platforms for platform in included_platforms):
```
Comment (mssalvatore) [inline generate_tags.py:70]: This is a lot of logic to digest all at once here. If we wrap it in a well-named function, for example `technique_matches_criteria()`, then we can just say
```python
if technique_matches_criteria(technique):
    # format and append
    ...
```

This way, If I want to understand what techniques are included, there's a clearly defined place where I can look. And if I don't care, I don't need to be destracted with those details while I'm reading the rest of the code.

NOTE: Naming is hard. `technique_matches_criteria` may not be the ideal name. The goal is to have something that reads as close to plain English as possible. `if technique_matches_criteria()` is pretty easy for anyone to understand because it's basically just english. Feel free to suggest another name.
